File: web/src/api/issue.ts
import type { User, Issue, IssueTemplate, Comment, Participant, MergeRequest, Review, ReviewStatus, ReviewComment, Label, Milestone } from '../lib/types'

type RequestFn = <T>(endpoint: string, options?: { method?: string; body?: any; token?: string | null }) => Promise<T>

export function createIssueApi(request: RequestFn, _baseUrl: string) {
  return {
    async listIssues(owner: string, repo: string) {
      return request<{ issues: Issue[]; participants: Participant[] }>(`/repos/${owner}/${repo}/issues`)
    },

    // Distinct issue/MR authors / assignees for filter dropdowns
    async listIssueAuthors(owner: string, repo: string, type?: 'issue' | 'mr') {
      const query = type === 'mr' ? '?type=mr' : ''
      return request<User[]>(`/repos/${owner}/${repo}/issues/authors${query}`)
    },

    async listIssueAssignees(owner: string, repo: string, type?: 'issue' | 'mr') {
      const query = type === 'mr' ? '?type=mr' : ''
      return request<User[]>(`/repos/${owner}/${repo}/issues/assignees${query}`)
    },

    async getIssue(owner: string, repo: string, issueId: number) {
      return request<Issue>(`/repos/${owner}/${repo}/issues/${issueId}`)
    },

    async updateIssue(owner: string, repo: string, issueId: number, data: { title?: string; body?: string; state?: string; milestoneId?: number | null }) {
      return request<Issue>(`/repos/${owner}/${repo}/issues/${issueId}`, {
        method: 'PATCH',
        body: data,
      })
    },

    async createIssue(owner: string, repo: string, title: string, content: string, assignees?: string[]) {
      return request<Issue>(`/repos/${owner}/${repo}/issues`, {
        method: 'POST',
        body: { title, body: content, assignees },
      })
    },

    // Issue assignees
    async listAssignees(owner: string, repo: string, issueId: number) {
      return request<{ assignees: User[] }>(`/repos/${owner}/${repo}/issues/${issueId}/assignees`)
    },

    async addAssignee(owner: string, repo: string, issueId: number, username: string) {
      return request<{ assignees: User[] }>(`/repos/${owner}/${repo}/issues/${issueId}/assignees/${username}`, {
        method: 'POST',
      })
    },

    async removeAssignee(owner: string, repo: string, issueId: number, username: string) {
      return request<{ assignees: User[] }>(`/repos/${owner}/${repo}/issues/${issueId}/assignees/${username}`, {
        method: 'DELETE',
      })
    },

    // Issue locking
    async lockIssue(owner: string, repo: string, issueId: number) {
      return request<{ message: string }>(`/repos/${owner}/${repo}/issues/${issueId}/lock`, {
        method: 'POST',
      })
    },

    async unlockIssue(owner: string, repo: string, issueId: number) {
      return request<{ message: string }>(`/repos/${owner}/${repo}/issues/${issueId}/lock`, {
        method: 'DELETE',
      })
    },

    // Issue templates
    async getIssueTemplates(owner: string, repo: string) {
      return request<IssueTemplate[]>(`/repos/${owner}/${repo}/issues/templates`)
    },

    // Batch update issues
    async batchUpdateIssues(owner: string, repo: string, data: {
      issueIds: number[]
      closed?: boolean
      milestoneId?: number
      priorityId?: number
      addLabelIds?: number[]
      removeLabelIds?: number[]
    }) {
      return request<{ message: string; updated: number }>(`/repos/${owner}/${repo}/issues/batch`, {
        method: 'POST',
        body: data,
      })
    },

    async listIssueComments(owner: string, repo: string, issueId: number) {
      return request<{ comments: Comment[]; participants: Participant[] }>(`/repos/${owner}/${repo}/issues/${issueId}/comments`)
    },

    async addIssueComment(owner: string, repo: string, issueId: number, content: string) {
      return request<Comment>(`/repos/${owner}/${repo}/issues/${issueId}/comments`, {
        method: 'POST',
        body: { body: content },
      })
    },

    async updateIssueComment(owner: string, repo: string, issueId: number, commentId: number, content: string) {
      return request<Comment>(`/repos/${owner}/${repo}/issues/${issueId}/comments/${commentId}`, {
        method: 'PATCH',
        body: { content },
      })
    },

    async deleteIssueComment(owner: string, repo: string, issueId: number, commentId: number) {
      return request<void>(`/repos/${owner}/${repo}/issues/${issueId}/comments/${commentId}`, {
        method: 'DELETE',
      })
    },

    // Labels
    async listLabels(owner: string, repo: string) {
      return request<Label[]>(`/repos/${owner}/${repo}/labels`)
    },

    async createLabel(owner: string, repo: string, name: string, color: string) {
      return request<Label>(`/repos/${owner}/${repo}/labels`, {
        method: 'POST',
        body: { name, color },
      })
    },

    async updateLabel(owner: string, repo: string, labelId: number, data: Partial<Label>) {
      return request<Label>(`/repos/${owner}/${repo}/labels/${labelId}`, {
        method: 'PATCH',
        body: data,
      })
    },

    async deleteLabel(owner: string, repo: string, labelId: number) {
      return request<void>(`/repos/${owner}/${repo}/labels/${labelId}`, {
        method: 'DELETE',
      })
    },

    async addLabelToIssue(owner: string, repo: string, issueId: number, labelId: number) {
      return request<Label>(`/repos/${owner}/${repo}/issues/${issueId}/labels/${labelId}`, {
        method: 'POST',
      })
    },

    async removeLabelFromIssue(owner: string, repo: string, issueId: number, labelId: number) {
      return request<void>(`/repos/${owner}/${repo}/issues/${issueId}/labels/${labelId}`, {
        method: 'DELETE',
      })
    },

    // Milestones
    async listMilestones(owner: string, repo: string) {
      return request<Milestone[]>(`/repos/${owner}/${repo}/milestones`)
    },

    async getMilestone(owner: string, repo: string, milestoneId: number) {
      return request<Milestone>(`/repos/${owner}/${repo}/milestones/${milestoneId}`)
    },

    async createMilestone(owner: string, repo: string, title: string, description: string, dueDate?: string) {
      return request<Milestone>(`/repos/${owner}/${repo}/milestones`, {
        method: 'POST',
        body: { title, description, dueDate },
      })
    },

    async updateMilestone(owner: string, repo: string, milestoneId: number, title: string, description: string, dueDate?: string) {
      return request<Milestone>(`/repos/${owner}/${repo}/milestones/${milestoneId}`, {
        method: 'PATCH',
        body: { title, description, dueDate },
      })
    },

    async deleteMilestone(owner: string, repo: string, milestoneId: number) {
      return request<void>(`/repos/${owner}/${repo}/milestones/${milestoneId}`, {
        method: 'DELETE',
      })
    },

    async closeMilestone(owner: string, repo: string, milestoneId: number) {
      return request<void>(`/repos/${owner}/${repo}/milestones/${milestoneId}/close`, {
        method: 'POST',
      })
    },

    async reopenMilestone(owner: string, repo: string, milestoneId: number) {
      return request<void>(`/repos/${owner}/${repo}/milestones/${milestoneId}/reopen`, {
        method: 'POST',
      })
    },

    // Merge Requests
    async listMergeRequests(owner: string, repo: string, state?: string) {
      const params = state ? `?state=${state}` : ''
      const resp = await request<{ mergeRequests: MergeRequest[]; participants: Participant[] }>(`/repos/${owner}/${repo}/merge-requests${params}`)
      return resp || { mergeRequests: [], participants: [] }
    },

    async getMergeRequest(owner: string, repo: string, mrId: number) {
      return request<MergeRequest>(`/repos/${owner}/${repo}/merge-requests/${mrId}`)
    },

    async createMergeRequest(
      owner: string,
      repo: string,
      title: string,
      body: string,
      head: string,
      base: string,
      headUser?: string,
      headRepo?: string
    ) {
      return request<MergeRequest>(`/repos/${owner}/${repo}/merge-requests`, {
        method: 'POST',
        body: { title, body, head, base, headUser, headRepo },
      })
    },

    async mergeMergeRequest(owner: string, repo: string, mrId: number, strategy: string, message?: string) {
      return request<{ message: string; mergeCommit: string; strategy: string }>(
        `/repos/${owner}/${repo}/merge-requests/${mrId}/merge`,
        {
          method: 'POST',
          body: { strategy, message },
        }
      )
    },

    async updateMergeRequest(owner: string, repo: string, mrId: number, data: { title?: string; body?: string; state?: string }) {
      return request<MergeRequest>(`/repos/${owner}/${repo}/merge-requests/${mrId}`, {
        method: 'PATCH',
        body: data,
      })
    },

    async listMergeRequestComments(owner: string, repo: string, mrId: number) {
      return request<{ comments: Comment[]; participants: Participant[] }>(`/repos/${owner}/${repo}/issues/${mrId}/comments`)
    },

    async addMergeRequestComment(owner: string, repo: string, mrId: number, content: string) {
      return request<Comment>(`/repos/${owner}/${repo}/issues/${mrId}/comments`, {
        method: 'POST',
        body: { body: content },
      })
    },

    // Reviews
    async listReviews(owner: string, repo: string, mrId: number) {
      return request<Review[]>(`/repos/${owner}/${repo}/merge-requests/${mrId}/reviews`)
    },

    async createReview(
      owner: string,
      repo: string,
      mrId: number,
      status: ReviewStatus,
      content?: string
    ) {
      return request<Review>(`/repos/${owner}/${repo}/merge-requests/${mrId}/reviews`, {
        method: 'POST',
        body: { status, content: content || null },
      })
    },

    async updateReview(
      owner: string,
      repo: string,
      mrId: number,
      reviewId: number,
      status: ReviewStatus,
      content?: string
    ) {
      return request<void>(`/repos/${owner}/${repo}/merge-requests/${mrId}/reviews/${reviewId}`, {
        method: 'PATCH',
        body: { status, content: content || null },
      })
    },

    async deleteReview(owner: string, repo: string, mrId: number, reviewId: number) {
      return request<void>(`/repos/${owner}/${repo}/merge-requests/${mrId}/reviews/${reviewId}`, {
        method: 'DELETE',
      })
    },

    async listReviewComments(owner: string, repo: string, mrId: number, reviewId: number) {
      return request<ReviewComment[]>(
        `/repos/${owner}/${repo}/merge-requests/${mrId}/reviews/${reviewId}/comments`
      )
    },

    async createReviewComment(
      owner: string,
      repo: string,
      mrId: number,
      reviewId: number,
      filePath: string,
      line: number,
      content: string
    ) {
      return request<ReviewComment>(
        `/repos/${owner}/${repo}/merge-requests/${mrId}/reviews/${reviewId}/comments`,
        {
          method: 'POST',
          body: { filePath, line, content },
        }
      )
    },

    async deleteReviewComment(
      owner: string,
      repo: string,
      mrId: number,
      reviewId: number,
      commentId: number
    ) {
      return request<void>(
        `/repos/${owner}/${repo}/merge-requests/${mrId}/reviews/${reviewId}/comments/${commentId}`,
        { method: 'DELETE' }
      )
    },

    async getAheadBehind(owner: string, repo: string, branch: string, base: string) {
      return request<{ ahead: number; behind: number }>(
        `/repos/${owner}/${repo}/branches/${encodeURIComponent(branch)}/ahead-behind?base=${encodeURIComponent(base)}`
      )
    },

    // Search issues with advanced query syntax (is:open, author:xxx, label:xxx, ...)
    async searchIssues(owner: string, repo: string, q: string, page = 1, limit = 25) {
      const params = new URLSearchParams({ q, page: String(page), limit: String(limit) })
      return request<{
        issues: Issue[]
        total: number
        limit: number
        offset: number
        participants: Participant[]
      }>(`/repos/${owner}/${repo}/issues/search?${params.toString()}`)
    },

    async compare(
      owner: string,
      repo: string,
      base: string,
      head: string,
      headUser?: string,
      headRepo?: string
    ) {
      let url = `/repos/${owner}/${repo}/compare/${encodeURIComponent(base)}...${encodeURIComponent(head)}`
      const params = new URLSearchParams()
      if (headUser) params.append('headUser', headUser)
      if (headRepo) params.append('headRepo', headRepo)
      const queryString = params.toString()
      if (queryString) url += `?${queryString}`

      return request<{
        commits: Array<{ id: string; message: string; author: string; timestamp: string }>
        fileChanges: Array<{ filename: string; additions: number; deletions: number; status: string }>
        baseCommit: string
        headCommit: string
      }>(url)
    },
  }
}

File: server/internal/router/issue.go
package router

import (
	"iforge/iforge/internal/container"

	"github.com/gofiber/fiber/v2"
)

// registerIssueRoutes registers issue routes
func registerIssueRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/issues", c.IssueHandler().ListIssues)
	api.Post("/repos/:owner/:repo/issues", c.IssueHandler().CreateIssue)
	api.Post("/repos/:owner/:repo/issues/batch", c.IssueHandler().BatchUpdateIssues)
	api.Get("/repos/:owner/:repo/issues/templates", c.IssueHandler().GetIssueTemplates)
	api.Get("/repos/:owner/:repo/issues/authors", c.IssueHandler().ListIssueAuthors)
	api.Get("/repos/:owner/:repo/issues/assignees", c.IssueHandler().ListIssueAssignees)
	api.Get("/repos/:owner/:repo/issues/search", c.SearchHandler().SearchIssues)
	api.Get("/repos/:owner/:repo/issues/:id", c.IssueHandler().GetIssue)
	api.Patch("/repos/:owner/:repo/issues/:id", c.IssueHandler().UpdateIssue)
	api.Delete("/repos/:owner/:repo/issues/:id", c.IssueHandler().DeleteIssue)
	api.Post("/repos/:owner/:repo/issues/:id/lock", c.IssueHandler().LockIssue)
	api.Delete("/repos/:owner/:repo/issues/:id/lock", c.IssueHandler().UnlockIssue)
	api.Post("/repos/:owner/:repo/issues/:id/comments", c.IssueHandler().CreateComment)
	api.Get("/repos/:owner/:repo/issues/:id/comments", c.IssueHandler().ListComments)
	api.Patch("/repos/:owner/:repo/issues/:id/comments/:commentId", c.IssueHandler().UpdateComment)
	api.Delete("/repos/:owner/:repo/issues/:id/comments/:commentId", c.IssueHandler().DeleteComment)
	api.Get("/repos/:owner/:repo/issues/:id/assignees", c.IssueHandler().ListAssignees)
	api.Post("/repos/:owner/:repo/issues/:id/assignees/:username", c.IssueHandler().AddAssignee)
	api.Delete("/repos/:owner/:repo/issues/:id/assignees/:username", c.IssueHandler().RemoveAssignee)
	api.Get("/search/issues", c.SearchHandler().SearchAllIssues)
}

// registerMergeRequestRoutes registers merge request routes
func registerMergeRequestRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/merge-requests", c.MergeRequestHandler().ListMergeRequests)
	api.Post("/repos/:owner/:repo/merge-requests", c.MergeRequestHandler().CreateMergeRequest)
	api.Get("/repos/:owner/:repo/compare/:basehead", c.MergeRequestHandler().Compare)
	api.Get("/repos/:owner/:repo/merge-requests/:id", c.MergeRequestHandler().GetMergeRequest)
	api.Patch("/repos/:owner/:repo/merge-requests/:id", c.MergeRequestHandler().UpdateMergeRequest)
	api.Post("/repos/:owner/:repo/merge-requests/:id/merge", c.MergeRequestHandler().MergeMergeRequest)
	api.Get("/repos/:owner/:repo/merge-requests/:id/diff", c.MergeRequestHandler().GetMergeRequestDiff)
	api.Get("/repos/:owner/:repo/branches/:branch/ahead-behind", c.MergeRequestHandler().GetAheadBehind)
}

// registerLabelRoutes registers label routes
func registerLabelRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/labels", c.LabelHandler().ListLabels)
	api.Post("/repos/:owner/:repo/labels", c.LabelHandler().CreateLabel)
	api.Patch("/repos/:owner/:repo/labels/:id", c.LabelHandler().UpdateLabel)
	api.Delete("/repos/:owner/:repo/labels/:id", c.LabelHandler().DeleteLabel)
	api.Get("/repos/:owner/:repo/issues/:id/labels", c.LabelHandler().GetIssueLabels)
	api.Post("/repos/:owner/:repo/issues/:id/labels/:labelId", c.LabelHandler().AddLabelToIssue)
	api.Delete("/repos/:owner/:repo/issues/:id/labels/:labelId", c.LabelHandler().RemoveLabelFromIssue)
}

// registerMilestoneRoutes registers milestone routes
func registerMilestoneRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/milestones", c.MilestoneHandler().ListMilestones)
	api.Post("/repos/:owner/:repo/milestones", c.MilestoneHandler().CreateMilestone)
	api.Get("/repos/:owner/:repo/milestones/:id", c.MilestoneHandler().GetMilestone)
	api.Patch("/repos/:owner/:repo/milestones/:id", c.MilestoneHandler().UpdateMilestone)
	api.Delete("/repos/:owner/:repo/milestones/:id", c.MilestoneHandler().DeleteMilestone)
	api.Post("/repos/:owner/:repo/milestones/:id/close", c.MilestoneHandler().CloseMilestone)
	api.Post("/repos/:owner/:repo/milestones/:id/reopen", c.MilestoneHandler().ReopenMilestone)
}

// registerCustomFieldRoutes registers custom field routes
func registerCustomFieldRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/custom_fields", c.CustomFieldHandler().ListCustomFields)
	api.Post("/repos/:owner/:repo/custom_fields", c.CustomFieldHandler().CreateCustomField)
	api.Get("/repos/:owner/:repo/custom_fields/:id", c.CustomFieldHandler().GetCustomField)
	api.Patch("/repos/:owner/:repo/custom_fields/:id", c.CustomFieldHandler().UpdateCustomField)
	api.Delete("/repos/:owner/:repo/custom_fields/:id", c.CustomFieldHandler().DeleteCustomField)
	api.Get("/repos/:owner/:repo/issues/:id/custom_fields", c.CustomFieldHandler().GetIssueCustomFieldValues)
	api.Post("/repos/:owner/:repo/issues/:id/custom_fields/:fieldId", c.CustomFieldHandler().SetIssueCustomFieldValue)
}

// registerPriorityRoutes registers priority routes
func registerPriorityRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/priorities", c.PriorityHandler().ListPriorities)
	api.Post("/repos/:owner/:repo/priorities", c.PriorityHandler().CreatePriority)
	// Static paths must be registered before /:id to avoid being captured by the parameter
	api.Post("/repos/:owner/:repo/priorities/reorder", c.PriorityHandler().ReorderPriorities)
	api.Get("/repos/:owner/:repo/priorities/default", c.PriorityHandler().GetDefaultPriority)
	api.Post("/repos/:owner/:repo/priorities/default", c.PriorityHandler().SetDefaultPriority)
	api.Get("/repos/:owner/:repo/priorities/:id", c.PriorityHandler().GetPriority)
	api.Patch("/repos/:owner/:repo/priorities/:id", c.PriorityHandler().UpdatePriority)
	api.Delete("/repos/:owner/:repo/priorities/:id", c.PriorityHandler().DeletePriority)
}

// registerReviewRoutes registers review routes
func registerReviewRoutes(api fiber.Router, c *container.Container) {
	api.Get("/repos/:owner/:repo/merge-requests/:id/reviews", c.ReviewHandler().ListReviews)
	api.Post("/repos/:owner/:repo/merge-requests/:id/reviews", c.ReviewHandler().CreateReview)
	api.Get("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId", c.ReviewHandler().GetReview)
	api.Patch("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId", c.ReviewHandler().UpdateReview)
	api.Delete("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId", c.ReviewHandler().DeleteReview)
	api.Get("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId/comments", c.ReviewHandler().ListReviewComments)
	api.Post("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId/comments", c.ReviewHandler().CreateReviewComment)
	api.Patch("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId/comments/:commentId", c.ReviewHandler().UpdateReviewComment)
	api.Delete("/repos/:owner/:repo/merge-requests/:id/reviews/:reviewId/comments/:commentId", c.ReviewHandler().DeleteReviewComment)
}
